### ISSUE TITLE
NextJs SSR Amplify, cold start takes up to 7 seconds

### ISSUE THREAD

**Issue by vincent38wargnier** [aws-amplify/amplify-js#10992]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

Next.js

### Amplify APIs

Authentication, GraphQL API, Storage

### Amplify Categories

auth, storage, function, api, hosting

### Environment information

<details>

```
# Put output below this line

System:
    OS: macOS 11.5.1
    CPU: (8) arm64 Apple M1
    Memory: 128.09 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.8.0 - /opt/homebrew/bin/node
    Yarn: 1.22.17 - /opt/homebrew/bin/yarn
    npm: 7.21.0 - /opt/homebrew/bin/npm
  Browsers:
    Chrome: 109.0.5414.119
    Safari: 14.1.2
  npmPackages:
    @ampproject/toolbox-optimizer:  undefined ()
    @aws-amplify/ui-react: ^4.3.2 => 4.3.2 
    @aws-amplify/ui-react-internal:  undefined ()
    @aws-sdk/client-s3: ^3.224.0 => 3.226.0 (3.6.1)
    @babel/core:  undefined ()
    @babel/runtime:  7.15.4 
    @edge-runtime/primitives:  2.0.0 
    @hapi/accept:  undefined ()
    @napi-rs/triples:  undefined ()
    @next/react-dev-overlay:  undefined ()
    @react-icons/all-files: ^4.1.0 => 4.1.0 
    @segment/ajv-human-errors:  undefined ()
    @vercel/nft:  undefined ()
    acorn:  undefined ()
    amphtml-validator:  undefined ()
    anser:  undefined ()
    arg:  undefined ()
    assert:  undefined ()
    async-retry:  undefined ()
    async-sema:  undefined ()
    autoprefixer: ^10.4.13 => 10.4.13 
    aws-amplify: ^5.0.8 => 5.0.8 
    babel-packages:  undefined ()
    blueimp-load-image: ^5.16.0 => 5.16.0 
    browserify-zlib:  undefined ()
    browserslist:  undefined ()
    buffer:  undefined ()
    bytes:  undefined ()
    chalk:  undefined ()
    ci-info:  undefined ()
    cli-select:  undefined ()
    comment-json:  undefined ()
    compress.js: ^1.2.2 => 1.2.2 
    compression:  undefined ()
    conf:  undefined ()
    constants-browserify:  undefined ()
    content-disposition:  undefined ()
    content-type:  undefined ()
    cookie:  undefined ()
    cross-spawn:  undefined ()
    crypto-browserify:  undefined ()
    css.escape:  undefined ()
    cssnano-simple:  undefined ()
    data-uri-to-buffer:  undefined ()
    debug:  undefined ()
    devalue:  undefined ()
    domain-browser:  undefined ()
    edge-runtime:  undefined ()
    events:  undefined ()
    find-cache-dir:  undefined ()
    find-up:  undefined ()
    fresh:  undefined ()
    get-orientation:  undefined ()
    glob:  undefined ()
    gzip-size:  undefined ()
    http-proxy:  undefined ()
    https-browserify:  undefined ()
    icss-utils:  undefined ()
    ignore-loader:  undefined ()
    image-size:  undefined ()
    is-animated:  undefined ()
    is-docker:  undefined ()
    is-wsl:  undefined ()
    jest-worker:  undefined ()
    json5:  undefined ()
    jsonwebtoken:  undefined ()
    leaflet: ^1.9.3 => 1.9.3 
    leaflet-defaulticon-compatibility: ^0.1.1 => 0.1.1 
    loader-utils:  undefined ()
    lodash.curry:  undefined ()
    lru-cache:  undefined ()
    micromatch:  undefined ()
    mini-css-extract-plugin:  undefined ()
    nanoid:  undefined ()
    native-url:  undefined ()
    neo-async:  undefined ()
    next: 13.0.4 => 13.0.4 
    node-fetch:  undefined ()
    node-html-parser:  undefined ()
    ora:  undefined ()
    os-browserify:  undefined ()
    p-limit:  undefined ()
    path-browserify:  undefined ()
    piexifjs: ^1.0.6 => 1.0.6 
    platform:  undefined ()
    postcss: ^8.4.19 => 8.4.19 (8.4.14)
    postcss-flexbugs-fixes:  undefined ()
    postcss-modules-extract-imports:  undefined ()
    postcss-modules-local-by-default:  undefined ()
    postcss-modules-scope:  undefined ()
    postcss-modules-values:  undefined ()
    postcss-preset-env:  undefined ()
    postcss-safe-parser:  undefined ()
    postcss-scss:  undefined ()
    postcss-value-parser:  undefined ()
    process:  undefined ()
    punycode:  undefined ()
    querystring-es3:  undefined ()
    raw-body:  undefined ()
    react: 18.2.0 => 18.2.0 (18.3.0-next-4bd245e9e-20221104)
    react-beautiful-dnd: ^13.1.1 => 13.1.1 
    react-dom: 18.2.0 => 18.2.0 (18.3.0-next-4bd245e9e-20221104)
    react-icons: ^4.7.1 => 4.7.1 
    react-is:  18.2.0 
    react-leaflet: ^4.2.0 => 4.2.0 
    react-refresh:  0.12.0 
    react-server-dom-webpack:  undefined ()
    react-slideshow-image: ^4.1.0 => 4.1.1 
    react-tooltip: ^4.5.1 => 4.5.1 
    regenerator-runtime:  0.13.4 
    sass-loader:  undefined ()
    scheduler:  undefined ()
    schema-utils:  undefined ()
    semver:  undefined ()
    send:  undefined ()
    setimmediate:  undefined ()
    shell-quote:  undefined ()
    source-map:  undefined ()
    stacktrace-parser:  undefined ()
    stream-browserify:  undefined ()
    stream-http:  undefined ()
    string-hash:  undefined ()
    string_decoder:  undefined ()
    strip-ansi:  undefined ()
    tailwind-merge: ^1.8.0 => 1.8.0 
    tailwindcss: ^3.2.4 => 3.2.4 
    tar:  undefined ()
    terser:  undefined ()
    text-table:  undefined ()
    timers-browserify:  undefined ()
    tty-browserify:  undefined ()
    ua-parser-js:  undefined ()
    undici:  undefined ()
    unistore:  undefined ()
    util:  undefined ()
    vm-browserify:  undefined ()
    watchpack:  undefined ()
    web-vitals:  undefined ()
    webpack:  undefined ()
    webpack-sources:  undefined ()
    ws:  undefined ()
  npmGlobalPackages:
    @aws-amplify/cli: 10.5.2
    @ionic/cli: 6.17.1
    npm: 7.21.0
    prebuild-install: 7.1.1
    yarn: 1.22.17
```

</details>

### Describe the bug

Hi,
I have created an AWS Amplify Web app with NextJs SSR.
Everything works fine except one huge issue : cold start takes up to 7 seconds to have server response.

I realized that if I remove all my amplify backend, the cold start drops down to less than a second. So I guess the problem comes from the AWS Amplify libraries.

I deployed a version of my website that doesn't call any getServerSideProps, but the latency stays around 7 seconds.

This is my amplify backend configuration :

```
{
  "auth": {
    "userPoolGroups": {
      "service": "Cognito-UserPool-Groups",
      "providerPlugin": "awscloudformation",
      "dependsOn": [
        {
          "category": "auth",
          "resourceName": "ressourceNamef1xxxxxxxxx",
          "attributes": [
            "UserPoolId",
            "AppClientIDWeb",
            "AppClientID",
            "IdentityPoolId"
          ]
        }
      ]
    },
    "ressourceNamef1xxxxxxxxx": {
      "service": "Cognito",
      "providerPlugin": "awscloudformation",
      "dependsOn": [],
      "customAuth": false,
      "frontendAuthConfig": {
        "socialProviders": [],
        "usernameAttributes": [
          "EMAIL"
        ],
        "signupAttributes": [
          "EMAIL"
        ],
        "passwordProtectionSettings": {
          "passwordPolicyMinLength": 8,
          "passwordPolicyCharacters": []
        },
        "mfaConfiguration": "OFF",
        "mfaTypes": [
          "SMS"
        ],
        "verificationMechanisms": [
          "EMAIL"
        ]
      }
    }
  },
  "api": {
    "ressourceName": {
      "service": "AppSync",
      "providerPlugin": "awscloudformation",
      "dependsOn": [
        {
          "category": "auth",
          "resourceName": "ressourceNamef1xxxxxxxxx",
          "attributes": [
            "UserPoolId"
          ]
        }
      ],
      "output": {
        "authConfig": {
          "defaultAuthentication": {
            "authenticationType": "AMAZON_COGNITO_USER_POOLS",
            "userPoolConfig": {
              "userPoolId": "authressourceNamef1xxxxxxxxx"
            }
          },
          "additionalAuthenticationProviders": [
            {
              "authenticationType": "API_KEY",
              "apiKeyConfig": {
                "apiKeyExpirationDays": 365,
                "apiKeyExpirationDate": "2023-12-10T15:07:46.361Z",
                "description": ""
              }
            },
            {
              "authenticationType": "AWS_IAM"
            }
          ]
        }
      }
    }
  },
  "storage": {
    "s338e9026f": {
      "service": "S3",
      "providerPlugin": "awscloudformation",
      "dependsOn": [
        {
          "category": "auth",
          "resourceName": "ressourceNamef1xxxxxxxxx",
          "attributes": [
            "UserPoolId"
          ]
        },
        {
          "category": "auth",
          "resourceName": "userPoolGroups",
          "attributes": [
            "adminGroupRole"
          ]
        },
        {
          "category": "auth",
          "resourceName": "userPoolGroups",
          "attributes": [
            "superAdminGroupRole"
          ]
        }
      ]
    }
  },
  "function": {
    "ressourceNameSendEmails": {
      "build": true,
      "providerPlugin": "awscloudformation",
      "service": "Lambda",
      "dependsOn": [
        {
          "category": "api",
          "resourceName": "ressourceName",
          "attributes": [
            "GraphQLAPIIdOutput",
            "GraphQLAPIEndpointOutput",
            "GraphQLAPIKeyOutput"
          ]
        }
      ]
    },
    "getgooglereviews": {
      "build": true,
      "providerPlugin": "awscloudformation",
      "service": "Lambda",
      "dependsOn": [
        {
          "category": "auth",
          "resourceName": "ressourceNamef1xxxxxxxxx",
          "attributes": [
            "UserPoolId"
          ]
        },
        {
          "category": "api",
          "resourceName": "ressourceName",
          "attributes": [
            "GraphQLAPIIdOutput",
            "GraphQLAPIEndpointOutput",
            "GraphQLAPIKeyOutput"
          ]
        }
      ]
    }
  }
}
```

Here is the way my NexJs App is initialized with amplify : 

```
import '../styles/globals.css'
import { ContextProvider } from '../src/components/context/Context';
import Layout from '../src/components/Layout';
import '@aws-amplify/ui-react/styles.css';
import { Amplify } from 'aws-amplify';
import awsExports from '../src/aws-exports';

function MyApp({ Component, pageProps }) {
  Amplify.configure(awsExports);
//Amplify.configure({ ...awsExports, ssr: true }); //==> This doesn't change latency. issue
  return (
    <ContextProvider>
      <Layout>
        <Component {...pageProps} />
      </Layout>
    </ContextProvider>
  )
}


export default MyApp
```

I read on several forums that Amplify is not able to provide law latency cold start, so I better move to Vercel or Netlify (which I tried and works super well) for example. 
Can you please confirm the opposite?
Sources : 
https://www.reddit.com/r/nextjs/comments/uamqbl/beware_of_nextjs_on_aws_amplify/

I also found some other solutions like keeping the lambda warm with a cron job. Is that a good idea?
https://www.reddit.com/r/nextjs/comments/ml4i2t/ssr_pages_cold_start_improve_performance/

Thank you.

### Expected behavior

I would like to have cold start no longer than 1 second, as it is on most of other cloud providers.

### Reproduction steps

amplify init
amplify add auth
amplify add api
amplify add storage
amplify add function
amplify add hosting
amplify push

### Code Snippet

```javascript
// Put your code below this line.

```

### Log output

<details>

```
// Put your logs below this line


```

</details>

### aws-exports.js

/* eslint-disable */
// WARNING: DO NOT EDIT. This file is automatically generated by AWS Amplify. It will be overwritten.

const awsmobile = {
    "aws_project_region": "eu-west-2",
    "aws_cognito_identity_pool_id": "eu-west-2:xxxxxxxxxxxxxxx",
    "aws_cognito_region": "eu-west-2",
    "aws_user_pools_id": "eu-west-2_xxxxxxxxxxxxxxx",
    "aws_user_pools_web_client_id": "xxxxxxxxxxxxxxx",
    "oauth": {},
    "aws_cognito_username_attributes": [
        "EMAIL"
    ],
    "aws_cognito_social_providers": [],
    "aws_cognito_signup_attributes": [
        "EMAIL"
    ],
    "aws_cognito_mfa_configuration": "OFF",
    "aws_cognito_mfa_types": [
        "SMS"
    ],
    "aws_cognito_password_protection_settings": {
        "passwordPolicyMinLength": 8,
        "passwordPolicyCharacters": []
    },
    "aws_cognito_verification_mechanisms": [
        "EMAIL"
    ],
    "aws_appsync_graphqlEndpoint": "https://xxxxxxxxxxxxxxx.amazonaws.com/graphql",
    "aws_appsync_region": "eu-west-2",
    "aws_appsync_authenticationType": "AMAZON_COGNITO_USER_POOLS",
    "aws_appsync_apiKey": "xxxxxxxxxxxxxxx",
    "aws_user_files_s3_bucket": "xxxxxxxxxxxxxxx",
    "aws_user_files_s3_bucket_region": "eu-west-2"
};

export default awsmobile;

**Comment by nadetastic** [comment]
Hi @vincent38wargnier thanks for opening this issue. This seems a bit related to #8918 but I have a few questions:

- Have you been able to deploy and get the same results with `ssr: false` on `Amplify.configure()`?
- Are you experiencing the same latency on localhost?

**Comment by vincent38wargnier** [comment]
> Hi @vincent38wargnier thanks for opening this issue. This seems a bit related to #8918 but I have a few questions:
> 
> * Have you been able to deploy and get the same results with `ssr: false` on `Amplify.configure()`?
> * Are you experiencing the same latency on localhost?

Hi @nadetastic, you're right it's related to this issue, but mine being much longer I was wondering if it was the exact same source of problem.

I don't have the problem in localhost, it's only in prod, during cold start. So if no user have been on the website for few minutes, the next one that queries it will have a 7 sec latency more or less. Once the cold start has been done, refreshing the page just takes max 1 sec, even if it's on another browser or other machine. 

I saw that a solution is to keep it warm, with this plugin for example https://github.com/juanjoDiaz/serverless-plugin-warmup but I guess it's going to increase the general costs. I was also trying to make a lambda that pings the website every 2 minutes to keep the server busy.

I think the time that takes the cold start might me related to some serverless functions called by amplify, because when I remove amplify from the project and keep all the rest of the front end, the cold start time drops down to 1 sec. 

So I guess there is some time consuming in the cold start, might be the auth verification or some other amplify stuff. That's a good question I would say :/

**Comment by JianWang2016** [comment]
I'd like to confirm I experienced the exact same issue as @vincent38wargnier.  7 seconds for the app to respond and another 3-5 seconds to load a page using client-side data fetch of about forty rows of data from dynamodb.  Afterward, I went to another computer to bring up the app, it took less than 1 sec.  
AWS amplify team, this issue, if unresolved, makes the app unsuitable for real-world use.   I'd greatly appreciate you give it a high priority.

**Comment by vincent38wargnier** [comment]
Hi guys,
To find out what's the problem, I decided to restart a project from zero and following the documentation here :
https://docs.amplify.aws/start/getting-started/installation/q/integration/next/

My approach here is to publish the app at any step and compare the evolution of the cold start time.

- So if we follow the documentation, the first step is creating a NextJs SSR app, and initialize a base configuration of amplify, adding a Post table in the graphql table. Nothing is done yet in front-end NextJs with Amplify.
<img width="1138" alt="image" src="https://user-images.githubusercontent.com/33725800/220064942-6b3be80e-dede-4dc2-93d0-6eb160b64aad.png">

Here is the result of the cold start : 
<img width="1179" alt="image" src="https://user-images.githubusercontent.com/33725800/220065029-1ea2b3e2-f6bb-47a4-a60a-9a1cf54d85c1.png">
cold start is already 1.78sec with nothing happening in the backend yet, so not very hopefull for the next steps.

- Next step is the following : 
https://docs.amplify.aws/start/getting-started/data-model/q/integration/next/#create-a-graphql-api-and-database

So more or less we add a page index that ask the user to login through Cognito and then the user can access to the list of posts.
here is the code : 
```
// pages/index.js
import { Authenticator } from '@aws-amplify/ui-react';
import { Amplify, API, Auth, withSSRContext } from 'aws-amplify';
import Head from 'next/head';
import awsExports from '../aws-exports';
import { createPost } from '../graphql/mutations';
import { listPosts } from '../graphql/queries';
import styles from '../styles/Home.module.css';

Amplify.configure({ ...awsExports, ssr: true });

export async function getServerSideProps({ req }) {
  const SSR = withSSRContext({ req });
  
  try {
    const response = await SSR.API.graphql({ query: listPosts, authMode: 'API_KEY' });
    return {
      props: {
        posts: response.data.listPosts.items,
      },
    };
  } catch (err) {
    console.log(err);
    return {
      props: {},
    };
  }
}

async function handleCreatePost(event) {
  event.preventDefault();

  const form = new FormData(event.target);

  try {
    const { data } = await API.graphql({
      authMode: 'AMAZON_COGNITO_USER_POOLS',
      query: createPost,
      variables: {
        input: {
          title: form.get('title'),
          content: form.get('content')
        }
      }
    });

    window.location.href = `/posts/${data.createPost.id}`;
  } catch ({ errors }) {
    console.error(...errors);
    throw new Error(errors[0].message);
  }
}

export default function Home({ posts = [] }) {
  return (
    <div className={styles.container}>
      <Head>
        <title>Amplify + Next.js</title>
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <main className={styles.main}>
        <h1 className={styles.title}>Amplify + Next.js</h1>

        <p className={styles.description}>
          <code className={styles.code}>{posts.length}</code>
          posts
        </p>

        <div className={styles.grid}>
          {posts.map((post) => (
            <a className={styles.card} href={`/posts/${post.id}`} key={post.id}>
              <h3>{post.title}</h3>
              <p>{post.content}</p>
            </a>
          ))}

          <div className={styles.card}>
            <h3 className={styles.title}>New Post</h3>

            <Authenticator>
              <form onSubmit={handleCreatePost}>
                <fieldset>
                  <legend>Title</legend>
                  <input
                    defaultValue={`Today, ${new Date().toLocaleTimeString()}`}
                    name="title"
                  />
                </fieldset>

                <fieldset>
                  <legend>Content</legend>
                  <textarea
                    defaultValue="I built an Amplify project with Next.js!"
                    name="content"
                  />
                </fieldset>

                <button>Create Post</button>
                <button type="button" onClick={() => Auth.signOut()}>
                  Sign out
                </button>
              </form>
            </Authenticator>
          </div>
        </div>
      </main>
    </div>
  );
}
```

Result of cold start after publishing :
<img width="658" alt="image" src="https://user-images.githubusercontent.com/33725800/220066434-be7ae4ae-850d-452a-b500-27e6c5c62f67.png">
Okay here we are already at almost 7 seconds of cold start.

So after following the documentation, cold start is already too high after second step, showing that Amplify NextJs SRR is not exploitable for a professional use, unless there is a trick to fix that.

If you want to check the project, everything is accessible here :
https://github.com/vincent38wargnier/test-amplify-cold-start-nextjs-ssr

@nadetastic do you have any idea if we can expect a fix from the Amplify team, and if so, how long should it take ?
Otherwise I will have to move on another technology sadly.

Thank you.

**Comment by JianWang2016** [comment]
@vincent38wargnier, I greatly appreciate your thorough analyses.  In this [issue ](https://github.com/aws-amplify/amplify-hosting/issues/3211), it was mentioned the cold start was less dramatic when using next 13 app folder.  It is still a couple of seconds.  I tried to move my component to app folder but had problems placing Amylify.configure in either layout.jsx or index.jsx.  My app from time to time complains about no credentials, but not all the time though.  No such problem when using the pages folder.  I am puzzled that amplify team has pretty much done nothing for the above issue for two months.

**Comment by nadetastic** [comment]
@vincent38wargnier @JianWang2016 thanks for the additional information on this issue, very much appreciated.

I've been able to experience the similar cold start times with `next@13` and `next@12` and I'm discussing this with the team - will follow up here soon

**Comment by gornostal** [comment]
Is there any workaround to this? I have the same issue too.
Thanks!

**Comment by simotissi** [comment]
Same problem here. About 20sec for the first load.

**Comment by Gere-y-mir** [comment]
+1

**Comment by nadetastic** [comment]
Hi @gornostal @simotissi - can you confirm if this is happening only after deploying to Hosting or also on localhost?

**Comment by marcantoineveilleux** [comment]
For me, it's only while hosting. Localhost is fine.

**Comment by vincent38wargnier** [comment]
I also thought it was on hosting only but I noticed if I restart my app after a while on localhost I have the same cold start unless everything is mocked

**Comment by simotissi** [comment]
Yes @nadetastic , it just seems like a problem on Hosting. In localhost it is definitely faster, even if not immediate

**Comment by dylan-westbury** [comment]
We have a super basic static next.js web app that doesn't even call `Amplify.configure` and it's taking **20 seconds** on first cold start load.

Here is link: https://feature-landingpage.d16t4qlpm725tl.amplifyapp.com/

<img width="1586" alt="Screenshot 2023-04-11 at 10 44 14 am" src="https://user-images.githubusercontent.com/16129043/231026580-356d6418-8259-4ff2-ba39-872aef11c2a2.png">

**Comment by nodeover** [comment]
I'm having the same problem too. I hope this gets resolved soon.

**Comment by up1-io** [comment]
faced with the same problem

**Comment by abepuentes** [comment]
So this has state the same from Feb 17th, does anyone try this builds on SST? just wondering if there's better cold starts time.

**Comment by nadetastic** [comment]
Related Hosting issue [#3211](https://github.com/aws-amplify/amplify-hosting/issues/3211)

**Comment by vincent38wargnier** [comment]
It appears that cold start issues are prevalent in Next.js. The Next.js blog provides some improvements for cold starts in version 13 [nextjs blog](https://nextjs.org/blog/next-13) : 
<img width="786" alt="image" src="https://user-images.githubusercontent.com/33725800/235293558-ad87fa53-ffc4-419e-ae59-5d214e57f0f4.png">
Apparently, when using getServerSideProps, it takes at least 1.8 seconds for a cold start.

I have rebuilt my backend using Nest.js, yet I still experience cold starts that seem to be related to the content of my frontend website and the utilization of getServerSideProps.

Additionally, the time taken by the Amplify backend for authentication further increases the cold start by a few seconds at a minimum.

In conclusion, I believe there is an incompatibility between using serverless with Next.js and an Amplify backend. Generally speaking, Next.js seems to be somewhat incompatible with serverless when getServerSideProps is needed.

Personally, I will be migrating my project to a microservice architecture and will wait until serverless technologies become more suitable for web NextJs requirements.

**Comment by iskandarovBakshi** [comment]
I'm running into the exact problem in Dev environment in Nextjs project, I use latest version of aws-amplify, I use only these imports:
`import { Amplify } from "@aws-amplify/core"`
`import { withSSRContext } from "aws-amplify"`

It's almost impossible to use HMR, 6813 modules( 80 seconds ) compiled in first start, every time I change page it takes 50 seconds or so, it's just because I imported 2 statements from aws

<img width="561" alt="image" src="https://github.com/aws-amplify/amplify-js/assets/15240881/4a282706-fcdf-4df3-9b5a-7ee3b6dfe273">
<img width="465" alt="image" src="https://github.com/aws-amplify/amplify-js/assets/15240881/f471d90b-2021-4260-b334-3fdc2a29aae3">

**Comment by punit1108** [comment]
Facing the same issues, Though moving to Vercel helps this situation a lot. I noticed that there is almost no cold start there.

**Comment by abepuentes** [comment]
As a follow up I did migrate my project to SST and its working great, so if anyone need this resolve today just follow the migrate from Vercel guide on SST website and add 2 warm functions in the sst.config file. This will be a new deployment so you can keep you amplify hosting up for comparison. Btw our bill from just building on amplify hosting it was not as cheap as I first thought.

**Comment by haruharuharuby** [comment]
I'm currently facing the problem and maybe would resolved as the workaround.

With assuming the cause of this comes from the AWS Lambda Cold Start that including in the Amplify Hosting.
In my case by configuring the Route 53 Health Check to ping constantly to the CloudFront domain.

Related Post:
https://tobinc.medium.com/speeding-up-aws-amplify-nextjs-first-render-cold-start-and-images-an-unexpected-result-36a416d69615

**Comment by haruharuharuby** [comment]
Hope to implementing Provisioned Concurrency inside LambdaEdge which is included in the Amplify Hosting.

**Comment by tmb5cg** [comment]
I'm having the same issue, and replicated it by removing the Amplify library. Without the Amplify library, the cold start load is still at least 2 seconds. With Amplify library, it's 6 seconds or more.

Given that any page load greater than 1 second or something leads to crazy drops in traffic, I can only imagine how this has hurt my app. I always thought it was something I was doing wrong, but I sorta blindly trusted Amplify.

Huge bummer. Right when I finally understand Amplify really well, comfortable to build anything with users / user data etc sorta complex from the ground up in a weekend, I'm thinking I've wasted my time and should've learned vercel.

does anyone know of any tutorials on migrating away from Amplify?

**Comment by abepuentes** [comment]
Hey @tmb5cg this comment hit home, I can totally relate. To me the smoother path has been using SST, two commands to deploy the site using theirs migrate from Vercel guide and for free. They have guides to build all the amplify services in a performant way but the warm function option added a couple of weeks ago give much better TTFB.

**Comment by tmb5cg** [comment]
> Hey @tmb5cg this comment hit home, I can totally relate. To me the smoother path has been using SST, two commands to deploy the site using theirs migrate from Vercel guide and for free. They have guides to build all the amplify services in a performant way but the warm function option added a couple of weeks ago give much better TTFB.

appreciate it, our shared condolences! 

in theory my site is a simple react/nextJS app that can deploy anywhere if I strip the amplify auth. the only thing I need amplify for is:

- user account creation
- user login (which is only needed on my main app as check, is the user authorized, if so, enable premium features)

And account creation is behind a Stripe check to confirm payment via webhook, boom then we hit the create user, which has nothing to do with amplify but for the api call to create user. 

Now that I think about it, this is as simple a migration as one could want.

one interesting thing. I am building a desktop app in Electron where I sorta taped on an Amplify backend built for a website that will never exist. it works perfectly and for a desktop app, i'm not at the mercy of amplify serving the pages, the Amplify backend actually works really nicely. 

Question: in theory I wonder how a NextJS app with the amplify package would perform deployed on Vercel, if it's Amplify's cold start that's causing all these problems. although at the core of our problem, is it not our app being "serverless" itself that is the issue? and vercel is serverless too? sorta confusing wording but hopefully what i'm trying to say makes sense. 

either way i'm going to try to drop Amplify. sad that this will waste a whole sunday or more purely as a test but, can't emphasize enough the absolute most important thing on a website is its first load time. I unconciously bail from sooo many websites if they're not served instantly! 7 seconds, what the hell!

**Comment by abepuentes** [comment]
@tmb5cg totally relate again, we do use amplify as a backend for our apps and works great. Give a try to SST, this should not take more than 10 min.

```
// Run the following command in your folder  
npx create-sst@latest
// Add the warmer functions to the config file sst.config.ts
// 50 warmer functions cost 1 dollar total on AWS, our project is small 2 is fine for us.
const site = new NextjsSite(stack, "site", {
        warm: 2,
      });
 // Deploy to AWS, first time take a while. --stage can be dev or prod or whatever
npx set deploy --stage prod
```

**Comment by abdallahshaban557** [comment]
Hello everyone, the Amplify Hosting team is currently working on improving the cold start times - we will provide updates when we have them on this issue.

**Comment by Tomekmularczyk** [comment]
Hey guys I experience the same problem when using Vercel deployments. This is related to #11030 

Here is a very basic repo reproducing the problem: https://github.com/Tomekmularczyk/amplify-next and here is the recording:


https://github.com/aws-amplify/amplify-js/assets/12544704/6807fc40-4f9a-4515-83c0-3d8faf2a07c7

Note that:
- I'm using the `app` directory
- I'm configuring Amplify for server components and client components
- to reproduce the issue page **has to be dynamic** that's why it is under [lang]
- if I remove `import "./amplifyConfigureClient"; import "./amplifyConfigureServer";` problem does not happen
- the cold start happens on the first page load after site has been deployed on Vercel or after some idle time ~1h

@abdallahshaban557 should I create separate issue as this is somewhat different setup than what has been discussed here and it doesn't matter if it is AWS or Vercel hosting - delay happens in both.

**Comment by tmb5cg** [comment]
Hi all,

So I migrated hosting for two of my Anplify apps to Vercel. They function identically as before and load in less than 1 second compared to 6-12 seconds previously.

My backend is still entirely Amplify and it's a great product for that. I just hardcoded my keys which I know isn't great but I only use one backend environment so each deploy, both to the frontend or backend, doesn't break anything. 

I'm not sure what the issue others encountered, if someone can explain how my keys would change (like as in aws-exports passed into Amplify.configure) because it works great now.

I can't imagine how much traffic / potential customers I missed out on but it's been a great learning process either way (the exact stat is like 40% of users bail after 3 seconds and I 100% believe that). 

I really like Vercel's dashboard too, it's like having Amplify Studio backend visualizer except it's Vercel studio for your frontend. And you can integrate page speed + more cool stuff, and learn a new platform at the same time. 

My understanding is that Amplify is incredible as an all in one backend builder, correct me if I'm wrong, but as an all in one platform it's really meant as a test proof of concept. So naturally splitting the hosting/frontend from the backend in the end I have a really great setup (let's see if I eat my words lol)

I am SO glad I found this thread, not sure how I stumbled on it because up until a week ago I was still dead convinced I was doing something wrong / created a totally dependency bloated project  (which I rebuilt from scratch like 3 or 4 times in new Amplify environments attempting to fix what I now know is unfixable). Just searching "amplify next js react very slow first time to load" on Google, you'd never find it's a very real thing.

**Comment by fnavarrodev** [comment]
Those who can not pay Vercel I highly recommend SST. We moved our next.js apps from Amplify to SST. https://docs.sst.dev/start/nextjs The option to warm the lambda is a configuration flag but can't be used with Lambda@Edge, that in our case works better using Edge location. Give it a try. I'm not the first here in this thread recommending it.

**Comment by JakubBatel** [comment]
Hi everyone!

First of all, I would like to thank everyone who commented here it was a tremendous help and moral boost that I am not alone in this.

I would like to add my insight and recommend [Netlify](https://docs.netlify.com/welcome/add-new-site/) if your budget is tight (can not afford Vercel) and you are too lazy to play around with SST. It works great, deployment is super easy and works out of the box without any cold starts. And the best thing is that their free plan can be actually used for commercial use.

So if you are like me and building a portfolio website or any sort of small website I believe that Netlify is the best option for you.

I had one issue which I fixed sort of quickly and that is with i18n I got `Too Many Redirects` because Netlify setups some redirects on its own and NextJS middleware also tried to redirect so it ended in an infinite loop redirecting from `{URL}` to `{URL}/en` and then the other way around. I fixed that by turning on the language prefix for the default language as well.

**Comment by nadetastic** [comment]
Hello everyone, quick update on this, the Amplify Hosting team is currently working on improving the cold start times and are tracking this work on [this issue in the Hosting repo](https://github.com/aws-amplify/amplify-hosting/issues/3211). We will continue to provide updates here but please feel free to subscribe to that issue for updates directly from the Hosting team.

**Comment by vishalobrizum** [comment]
Hello, I followed the instructions and redeployed but I am seeing 15 seconds delay 😢 . At this rate, this is not usable in production. I am using Amplify to host the app. It's a nextjs 13 app directory app:

<img width="1135" alt="Screenshot 2023-08-09 at 8 05 36 PM" src="https://github.com/aws-amplify/amplify-js/assets/125282700/bb4f0666-13ef-411a-b70d-149efe014beb">

Any help would be greatly appreciated!!

**Comment by ConorFl** [comment]
Still seeing significant cold start delays (aws-amplify 5.3.8, next 13.4.16). I was able to solve it with the health check suggested above, but paying a dollar month to make my code work is a bummer.

**Comment by adarsh-drishya** [comment]
hi everyone , https://nextjs.org/docs/app/api-reference/next-config-js/optimizePackageImports use it, i saw significant improvement

**Comment by Tomekmularczyk** [comment]
> hi everyone , https://nextjs.org/docs/app/api-reference/next-config-js/optimizePackageImports use it, i saw significant improvement

can you share the exact config you use?

**Comment by adarsh-drishya** [comment]
first of all it is released yesterday @Tomekmularczyk so just update your nextjs version to 13.5.2 , then do the above

**Comment by aprilmintacpineda** [comment]
Closed even though not fixed. Typical move of AWS Amplify. That's why I stopped using this tool. They really cannot fix this because using docker for aws lambda will inevitably increase your cold-start. The only solution is to not use docker.

**Comment by kitdesai** [comment]
why was this closed? this is definitely NOT fixed. I am still experiencing these issues with Amplify and my Next.js app running the latest version. I guess I need to try Vercel or SST instead?

**Comment by tmb5cg** [comment]
> why was this closed? this is definitely NOT fixed. I am still experiencing these issues with Amplify and my Next.js app running the latest version. I guess I need to try Vercel or SST instead? 

The longer you wait the more customers you're losing lol

**Comment by vincent38wargnier** [comment]
> why was this closed? this is definitely NOT fixed. I am still experiencing these issues with Amplify and my Next.js app running the latest version. I guess I need to try Vercel or SST instead?

I gave up wih serverless for backend. I'm using classical micro service, and Vercel for Serverless Front. And even like that I have like 3sec cold start, cause you need to optimize your front as f*ck to reduce cold start. But cold start is definitly shorter on Vercel

**Comment by RoseSecurity** [comment]
Any way we could reopen this ticket? My Next.js app is crying out in pain from how cold the start is

```console
➜  ~ curl -s --connect-timeout 5 -w "time_namelookup: %{time_namelookup}, time_connect: %{time_connect},
time_appconnect: %{time_appconnect}, time_pretransfer: %{time_pretransfer}, time_redirect: %{time_redirect},
time_starttransfer: %{time_starttransfer}, 
time_total: %{time_total}\n" -o /dev/null https://whywasthisticketclosed.com

time_namelookup: 0.195355
time_connect: 0.223631
time_appconnect: 0.266611
time_pretransfer: 0.266858
time_redirect: 0.000000
time_starttransfer: 11.731117
time_total: 11.740227
```

**Comment by JakubBatel** [comment]
I am glad I decided to switch to Netlify (6 months ago). From time to time I get an email about updates on this issue. So I check it out just out of curiosity and I have to say that I don't regret ditching the AWS Amplify as even a half year later the most basic functionality like loading a website in less then 4 seconds is still impossible. This is miles away to be production ready.
